### PR TITLE
Use standard dependency workflow

### LINKED ISSUE #55: Implement Dependency RFC 0004 for  composer
Implement [Dependency Management RFC Phase 1](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md) for `composer`. Check out the RFC for more details and background.

### 1. Determine dependency source strategy
<details>
<summary>Background</summary>

When possible, dependencies should be used directly from upstream, rather than undergoing any additional compilation or modifications performed by Paketo-maintained code. For each dependency, the corresponding buildpack maintainer group will decide if the dependency can be used directly from upstream, and must identify the location from which the dependency will be pulled from. Some of the Paketo Java buildpacks perform directory stripping during the buildpack build process itself. This could be a viable alternative to performing directory modifications during the dependency management process for maintainers to consider.
</details>

- [ ] Determine whether the dependency can be used directly from its upstream, rather than undergoing additional compilation/modification. 
* If using directly from upstream, note where dependencies will come from
* If compiled or modified, note the decision for this
- [ ] Document decision in a language-family level RFC. This decision can be combined in the RFC with the decision made for other buildpacks in the language family.

### 2. Version retrieval and metadata generation code
Please refer to the [retrieval RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#1-version-retrieval-make-retrieve) on version retrieval for details around the API and background.

- [ ] Create retrieval code for the dependency
* Will live in the buildpack under: `dependency/retrieval/`
* Per [the RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#transfercombine-version-retrieval-and-metadata-generation), this will be a combination of (1) discovering new dependency versions based on the `buildpack.toml`, and (2) generating metadata for each new version.
* **Note** : Per [the RFC caveat](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#caveat-compiled-dependencies) - if the dependency is to be compiled, the `SHA256` and `URI` field from the metadata should be omitted in this step.

### 3. Compilation code (if needed)
If the dependency will be compiled/modified, then refer to the [compilation RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#2-compilation) for API details and background. 

- [ ] Create compilation action
* Will live in the buildpack under: `dependency/actions/compile/`

### 4. Dependency testing (optional)
It's up to maintainer discretion if the dependency will be tested. It's recommended to test the dependency if it's been compiled. Check out the [testing RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#3-test-dependency-make-test) for details.

- [ ] Add tests for dependency
* Will live in the buildpack under: `dependency/test.`

### 5. Makefile Setup
When using the generalized workflows for dependency management down the line, version retrieval and dependency testing will be executed via a Makefile in order to provide the workflow a standardized way to run the code, regardless of what language it was written in. Check out the [RFC section](https://github.com/paketo-buildpacks/rfcs/blob/main/text/dependencies/rfcs/0004-dependency-management-phase-one.md#overarching-plan) for what this should look like.

- [ ] Add Makefile
* Will live in the buildpack at: `dependency/Makefile`

### 6. Leveraging the new code
This issue serves to set up all the main logic for the dependency management. The work to actually leverage this code and migrate off of the dep-server will be completed in a separate issue once workflows and infrastructure is set up.

### PR DESCRIPTION
**Pull request by joshuatcasey**
Use standard dependency workflow

Resolves #55 

Note that this includes ASC verification which did not exist in `dep-server` (https://github.com/joshuatcasey/dep-server/blob/main/pkg/dependency/composer.go)

### CODE REVIEW

**Comment by joshuatcasey** [comment]
BLOCKED on https://github.com/paketo-buildpacks/rfcs/pull/257

File: .github/workflows/update-dependencies-from-metadata.yml
Comment (thitch97) [inline .github/workflows/update-dependencies-from-metadata.yml:55]: I wonder if it makes sense to switch to the [environment files](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#environment-files) paradigm for setting output now rather than later, especially since these are new workflows. GHAs won't fully remove `set-output` support until May 2023 so I can see the argument for leaving things as is for a later concerted effort, but seems like fairly low-hanging fruit.

**Comment by sophiewigmore** [comment]
tested this code with the validator and the retrieval step seems reasonable. It outputs metadata that is properly formatted. I notice there is no license information for it. 

I'm assuming there is no license information?


Also, I am just wondering if you got a chance to test the dependency on the jammy stack. Did it work? I'm assuming we will use the same dependency across all stacks since the stack metadata is `*`, but double checking

File: .github/workflows/update-dependencies-from-metadata.yml
Comment (sophiewigmore) [reply .github/workflows/update-dependencies-from-metadata.yml:55]: If we do this, lets make the change in the upstream github-config workflow!

**Comment by sophiewigmore** [comment]
Also, do you think it would be worthwhile to add some basic tests for the code you have in the `main.go` retrieval code to make sure everything is working as expected?

**Comment by sophiewigmore** [comment]
Since we're moving to a checksum field, it may be wise to add support for checksum field cache layer checks during build. This change can also be made later once the first batch of dependency updates comes out, but just wanted to point that out.

File: .github/workflows/update-dependencies-from-metadata.yml
Comment (joshuatcasey) [reply .github/workflows/update-dependencies-from-metadata.yml:55]: Agreed, let's move the discussion to `github-config` as part of https://github.com/paketo-buildpacks/github-config/pull/604

**Comment by joshuatcasey** [comment]
> tested this code with the validator and the retrieval step seems reasonable. It outputs metadata that is properly formatted. I notice there is no license information for it.
> 
> I'm assuming there is no license information?

I didn't investigate the license information. The current `composer` dependency also has no license information and the proposed workflow generates no license information.

That being said, there's clearly a license associated with PHP Composer (`MIT License`? https://github.com/composer/composer/blob/main/LICENSE) so perhaps we should hardcode that?

**Comment by joshuatcasey** [comment]
> Since we're moving to a checksum field, it may be wise to add support for checksum field cache layer checks during build. This change can also be made later once the first batch of dependency updates comes out, but just wanted to point that out. Integration tests will fail on dependency bumps until this change is made.

Agreed. I prefer to keep this PR focused on the dependency workflow. The `sha256=>checksum` change can be made at the point it's needed, which is when the new dependencies are added.

**Comment by joshuatcasey** [comment]
> Also, do you think it would be worthwhile to add some basic tests for the code you have in the `main.go` retrieval code to make sure everything is working as expected?

I did think about this. I agree that the metadata generation (and checksum+signature verification) code is messy, but I wasn't 100% sure in this case that testing would add to my confidence that it's working correctly.

For example, I'm a lot more worried about not detecting new versions and having the workflow ✅  even though it's not picking up the new versions. For that scenario, this retrieval code is delegating to `libdependency` which I feel is pretty well tested.

If something does go wrong when generating metadata, an issue will soon be filed against the repo so at least a human can intervene at that time (https://github.com/paketo-buildpacks/github-config/issues/592).

**Comment by sophiewigmore** [comment]
RE licenses - I think we should investigate why nothing is found. I am weary of hardcoding something

**Comment by sophiewigmore** [comment]
> I did think about this. I agree that the metadata generation (and checksum+signature verification) code is messy, but I wasn't 100% sure in this case that testing would add to my confidence that it's working correctly.

A big part of moving off of the dep-server into the new code here is to reduce tech debt and make stuff easier for us to maintain. If the code is "messy" then I do think it would benefit from some tests to make sure we're doing the right thing. Relying on the workflow alone could be error-prone. The workflow is super complicated and may fail in weird places if.
For example, if the metadata generated is malformed in some way like missing a field or incorrect I'm not sure our workflows would catch that.

I think it would be valuable to have some kind of test for checking metadata generation is working as expected at the very least.

Also, having a test to make sure version discovery works properly isn't a terrible idea. It could fail silently with no GHA failures if no versions are picked up.

The bits that heavily rely on libdependency code I don't feel like need to be tested because you're right they've been tested thoroughly.

That's my opinion on this, I do feel like any code we heavily rely on in our automation should be tested. I notice other dependencies aren't tested (passenger buildpack) so maybe worth aligning on. @thitch97 do you have strong opinions either way as a maintainer of this code?

**Comment by joshuatcasey** [comment]
> RE licenses - I think we should investigate why nothing is found. I am weary of hardcoding something

Good idea. Turns out that the download archive isn't a tarball or a zip but a custom `phar` which can be extracted via `phar extract -f`, and `phar is available on the github virtual environments.

So it turns out that we [can retrieve the licenses](https://github.com/paketo-buildpacks/dep-server/blob/88f88eec065af9877656d2c1a314f64a3eab2437/pkg/dependency/licenses/licenses.go#L23-L29), and it correctly finds `[MIT]`.

**Comment by joshuatcasey** [comment]
@sophiewigmore I've added "some" tests. It's not exhaustive, but let's chat before I spend more time on it.

File: dependency/retrieval/main_test.go
Comment (sophiewigmore) [inline dependency/retrieval/main_test.go:78]: What's the point of this function?

**Comment by sophiewigmore** [comment]
@joshuatcasey the tests are on the right track 👍

File: dependency/retrieval/main_test.go
Comment (joshuatcasey) [reply dependency/retrieval/main_test.go:78]: `phar extract -f` is really verbose. This `Example` function demonstrates that it logs nothing to `STDOUT` on success.

I tried to implement a negative version of this (showing that `phar extract -f` output was sent to `STDOUT` on failure) but it contained dynamic output and was therefore not workable as an `Example` test. It could be tested as a unit test, but that would require injecting an `io.Writer` and a bunch of other changes that I wasn't prepared to do.

Comment (sophiewigmore) [reply dependency/retrieval/main_test.go:78]: I see. Are you using it anywhere? I don't think so but might have missed it. If not, I don't think we need this here. A comment will suffice with the behaviour of `phar extract -f`

Comment (joshuatcasey) [reply dependency/retrieval/main_test.go:78]: It's an [example test](https://go.dev/blog/examples) so it's run as part of `go test`.

Note that this does not apply when using [`go test -run Unit`](https://github.com/paketo-buildpacks/github-config/blob/main/implementation/scripts/unit.sh#L53), which is why the Library `./scripts/unit.sh` includes [`go test -run Unit|Example`](https://github.com/paketo-buildpacks/github-config/blob/main/library/scripts/unit.sh#L53).

Comment (sophiewigmore) [reply dependency/retrieval/main_test.go:78]: Cool, I haven't seen that before. Thanks for the explainer.

File: dependency/retrieval/main_test.go
Comment (joshuatcasey) [reply dependency/retrieval/main_test.go:99]: It's a required as part of the `Example` test. It's confusing but indicates "no output".

Comment (sophiewigmore) [reply dependency/retrieval/main_test.go:99]: Thanks!!

**Review by sophiewigmore — DISMISSED** [review]

**Review by sophiewigmore — APPROVED** [review]

**Pull request merged**
